storage/innobase/clone/clone0clone.cc

```diff
@@ -0,0 +1,2395 @@
+        return (0);
+      }
+    } else if (free_idx == SNAPSHOT_ARR_SIZE) {
+      free_idx = idx;
+    }
+  }
+
+  if (free_idx == SNAPSHOT_ARR_SIZE ||
+      (hdl_type == CLONE_HDL_COPY && m_num_snapshots == MAX_SNAPSHOTS) ||
+      (hdl_type == CLONE_HDL_APPLY && m_num_apply_snapshots == MAX_SNAPSHOTS)) {
+    my_error(ER_CLONE_TOO_MANY_CONCURRENT_CLONES, MYF(0), MAX_SNAPSHOTS);
+    return (ER_CLONE_TOO_MANY_CONCURRENT_CLONES);
+  }
+
+  /* Create a new snapshot. */
+  snapshot = UT_NEW(Clone_Snapshot(hdl_type, clone_type, free_idx, snapshot_id),
+                    mem_key_clone);
+
+  if (snapshot == nullptr) {
+    my_error(ER_OUTOFMEMORY, MYF(0), sizeof(Clone_Snapshot));
+    return (ER_OUTOFMEMORY);
+  }
+
+  m_snapshot_arr[free_idx] = snapshot;
+
+  if (hdl_type == CLONE_HDL_COPY) {
+    ++m_num_snapshots;
+  } else {
+    ut_ad(hdl_type == CLONE_HDL_APPLY);
+    ++m_num_apply_snapshots;
+  }
+
+  snapshot->attach(hdl_type, is_pfs_monitor);
+
+  return (0);
+}
+
+void Clone_Sys::detach_snapshot(Clone_Snapshot *snapshot,
+                                Clone_Handle_Type hdl_type) {
+  mysql_mutex_assert_owner(&m_clone_sys_mutex);
+  snapshot->detach();
+
+  /* Drop the snapshot. */
+  uint index;
+
+  index = snapshot->get_index();
+  ut_ad(m_snapshot_arr[index] == snapshot);
+
+  UT_DELETE(snapshot);
+
+  m_snapshot_arr[index] = nullptr;
+
+  if (hdl_type == CLONE_HDL_COPY) {
+    ut_ad(m_num_snapshots > 0);
+    --m_num_snapshots;
+
+  } else {
+    ut_ad(hdl_type == CLONE_HDL_APPLY);
+    ut_ad(m_num_apply_snapshots > 0);
+    --m_num_apply_snapshots;
+  }
+}
+
+Clone_Sys::Acquire_clone::Acquire_clone() {
+  std::tie(std::ignore, m_clone) = clone_sys->check_active_clone();
+
+  if (m_clone != nullptr) {
+    m_clone->attach();
+  }
+}
+
+Clone_Sys::Acquire_clone::~Acquire_clone() {
+  if (m_clone != nullptr) {
+    clone_sys->drop_clone(m_clone);
+  }
+  m_clone = nullptr;
+}
+
+Clone_Snapshot *Clone_Sys::Acquire_clone::get_snapshot() {
+  if (m_clone == nullptr) {
+    return nullptr; /* purecov: inspected */
+  }
+  return m_clone->get_snapshot();
+}
+
+bool Clone_Sys::check_active_clone(bool print_alert) {
+  bool active_clone = false;
+  std::tie(active_clone, std::ignore) = check_active_clone();
+
+  if (active_clone && print_alert) {
+    /* purecov: begin inspected */
+    ib::info() << "DDL waiting for CLONE to abort";
+    /* purecov: end */
+  }
+  return (active_clone);
+}
+
+std::tuple<bool, Clone_Handle *> Clone_Sys::check_active_clone() {
+  mysql_mutex_assert_owner(&m_clone_sys_mutex);
+
+  bool active_clone = false;
+  Clone_Handle *active_handle = nullptr;
+
+  /* Check for active clone operations. */
+  for (int idx = 0; idx < CLONE_ARR_SIZE; idx++) {
+    auto clone_hdl = m_clone_arr[idx];
+
+    if (clone_hdl != nullptr && clone_hdl->is_copy_clone()) {
+      active_clone = true;
+      active_handle = clone_hdl;
+      break;
+    }
+  }
+  return std::make_tuple(active_clone, active_handle);
+}
```

Comment:
The first component of the tuple is completely redundant, and it will incur runtime overhead. Please let us just return a pointer:
```c++
Clone_handle *Clone_sys::check_active_clone()
{
  mysql_mutex_assert_owner(&m_clone_sys_mutex);
  for (Clone_handle h: m_clone_arr)
    if (h && h->is_copy_clone())
      return h;
  return nullptr;
}
```